<h1>Question 3</h1>

<b>3. Feed the following paragraph into your favourite data analytics tool, and answer the following;

a. What is the probability of the word “data” occurring in each line ?

b. What is the distribution of distinct word counts across all the lines ?

c. What is the probability of the word “analytics” occurring after the word “data” ?
    
==================================================================================
    
As a term, data analytics predominantly refers to an assortment of applications, from basic business intelligence (BI), reporting and online analytical processing (OLAP) to various forms of advanced analytics. In that sense, it's similar in nature to business analytics, another umbrella term for approaches to analyzing data -- with the difference that the latter is oriented to business uses, while data analytics has a broader focus. The expansive view of the term isn't universal, though: In some cases, people use data analytics specifically to mean advanced analytics, treating BI as a separate category. Data analytics initiatives can help businesses increase revenues, improve operational efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of boosting business performance. Depending on the particular application, the data that's analyzed can consist of either historical records or new information that has been processed for real-time analytics uses. In addition, it can come from a mix of internal systems and external data sources. At a high level, data analytics methodologies include exploratory data analysis (EDA), which aims to find patterns and relationships in data, and confirmatory data analysis (CDA), which applies statistical techniques to determine whether hypotheses about a data set are true or false. EDA is often compared to detective work, while CDA is akin to the work of a judge or jury during a court trial -- a distinction first drawn by statistician John W. Tukey in his 1977 book Exploratory Data Analysis. Data analytics can also be separated into quantitative data analysis and qualitative data analysis. The former involves analysis of numerical data with quantifiable variables that can be compared or measured statistically. The qualitative approach is more interpretive -- it focuses on understanding the content of non-numerical data like text, images, audio and video, including common phrases, themes and points of view.</b>

<b>a. What is the probability of the word “data” occurring in each line ?</b>

In [14]:
import re
import pandas as pd
import nltk
from collections import Counter
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
#nltk.download('stopwords')
#nltk.download("punkt")

In [15]:
def tokenize(string):
    return re.compile('\w+').findall(string)

from collections import Counter

def word_freq(string): 
    text = tokenize(string.lower())
    c = Counter(text)           # count the words
    d = Counter(''.join(text))  # count all letters
    return (dict(c),dict(d))    # return a tuple of counted words and letters


data = "As a term, data analytics predominantly refers to an assortment of applications, from basic business intelligence (BI), reporting and online analytical processing (OLAP) to various forms of advanced analytics. In that sense, it's similar in nature to business analytics, another umbrella term for approaches to analyzing data -- with the difference that the latter is oriented to business uses, while data analytics has a broader focus. The expansive view of the term isn't universal, though: In some cases, people use data analytics specifically to mean advanced analytics, treating BI as a separate category. Data analytics initiatives can help businesses increase revenues, improve operational efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of boosting business performance. Depending on the particular application, the data that's analyzed can consist of either historical records or new information that has been processed for real-time analytics uses. In addition, it can come from a mix of internal systems and external data sources. At a high level, data analytics methodologies include exploratory data analysis (EDA), which aims to find patterns and relationships in data, and confirmatory data analysis (CDA), which applies statistical techniques to determine whether hypotheses about a data set are true or false. EDA is often compared to detective work, while CDA is akin to the work of a judge or jury during a court trial -- a distinction first drawn by statistician John W. Tukey in his 1977 book Exploratory Data Analysis. Data analytics can also be separated into quantitative data analysis and qualitative data analysis. The former involves analysis of numerical data with quantifiable variables that can be compared or measured statistically. The qualitative approach is more interpretive -- it focuses on understanding the content of non-numerical data like text, images, audio and video, including common phrases, themes and points of view."

words, letters = word_freq(data) # count and get dicts with counts

sumWords = sum(words.values())       # sum total words
sumLetters = sum(letters.values())   # sum total letters

# calc / print probability of word
for w in words:
    if w == "data":
        print("Probability of '{}': {}".format(w,words[w]/sumWords))

Probability of 'data': 0.05625


Probability of the word 'data' is 0.05625.


<b>b. What is the distribution of distinct word counts across all the lines ?</b>

In [16]:
text_tokens = word_tokenize(data.lower())

# generate list without stopwords and punctuation
text_tokens = [word for word in text_tokens if word.isalnum()]
tokens_without_sw = [word for word in text_tokens if not word in stopwords.words()]

In [17]:
# compute word frequency distribution across all the lines
pd.set_option('display.max_rows', None)
fdist = nltk.FreqDist(tokens_without_sw)
df_fdist = pd.DataFrame.from_dict(fdist, orient='index')
df_fdist.columns = ['Count']
df_fdist.index.name = 'Word'
df_fdist.sort_values('Count', ascending=False)

Count
Word                
data              18
analytics         10
analysis           6
business           4
term               3
view               2
qualitative        2
uses               2
eda                2
advanced           2
exploratory        2
cda                2
compared           2
work               2
applies            1
set                1
whether            1
true               1
false              1
often              1
determine          1
techniques         1
detective          1
statistical        1
hypotheses         1
patterns           1
confirmatory       1
relationships      1
judge              1
find               1
aims               1
include            1
methodologies      1
level              1
high               1
sources            1
external           1
systems            1
internal           1
mix                1
addition           1
akin               1
trial              1
jury               1
text               1
approach           1
interpretive       1
focuses            1
understanding      1
content            1
like               1
images             1
measured           1
audio              1
video              1
including          1
common             1
phrases            1
themes             1
statistically      1
variables          1
court              1
tukey              1
information        1
distinction        1
first              1
drawn              1
statistician       1
john               1
1977               1
quantifiable       1
book               1
separated          1
quantitative       1
former             1
involves           1
numerical          1
processed          1
either             1
new                1
records            1
approaches         1
analyzing          1
difference         1
latter             1
oriented           1
broader            1
focus              1
expansive          1
universal          1
though             1
cases              1
people             1
use                1
specifically       1
mean               1
umbrella           1
another            1
nature             1
reporting          1
predominantly      1
refers             1
assortment         1
applications       1
basic              1
intelligence       1
online             1
similar            1
analytical         1
processing         1
olap               1
various            1
forms              1
sense              1
treating           1
separate           1
category           1
boosting           1
gain               1
competitive        1
edge               1
rivals             1
ultimate           1
goal               1
performance        1
market             1
depending          1
particular         1
application        1
analyzed           1
consist            1
historical         1
trends             1
emerging           1
initiatives        1
efficiency         1
help               1
businesses         1
increase           1
revenues           1
improve            1
operational        1
optimize           1
quickly            1
marketing          1
campaigns          1
customer           1
service            1
efforts            1
respond            1
points             1

<b>c. What is the probability of the word “analytics” occurring after the word “data” ?</b>

In [21]:
bigramwords = re.findall('\w+', data.lower())
bigram_counts = Counter(zip(bigramwords,bigramwords[1:]))
data_analytics_count = bigram_counts[("data", "analytics")]
print(f"Number of times 'data analytics' occur together: {data_analytics_count}")
print(f"Number of times the word 'analytics' occur: {fdist['analytics']}")
print(f"Probablity of the word 'analytics' appearing after the word 'data': {data_analytics_count/fdist['analytics']}")

Number of times 'data analytics' occur together: 6
Number of times the word 'analytics' occur: 10
Probablity of the word 'analytics' appearing after the word 'data': 0.6


The probability of the word “analytics” occurring after the word “data” is 0.6.